js/markdown-stream/markdown-stream.ts

```diff
@@ -0,0 +1,257 @@
+        setTimeout(
+          () => btn.classList.remove("code-copy-button-checked"),
+          2000
+        );
+        e.clearSelection();
+      });
+    });
+  }
+
+  // ------- Scrolling logic -------
+
+  // Nearest scrollable parent element (if any)
+  #scrollableElement: HTMLElement | null = null;
+  // Whether content is currently being added to the element
+  #isContentBeingAdded = false;
+  // Whether the user has scrolled away from the bottom
+  #isUserScrolled = false;
+
+  #onScroll = (): void => {
+    if (!this.#isContentBeingAdded) {
+      this.#isUserScrolled = !this.#isNearBottom();
+    }
+  };
+
+  #isNearBottom(): boolean {
+    const el = this.#scrollableElement;
+    if (!el) return false;
+
+    return el.scrollHeight - (el.scrollTop + el.clientHeight) < 50;
+  }
+
+  #updateScrollableElement(): void {
+    const el = this.#findScrollableParent();
+
+    if (el !== this.#scrollableElement) {
+      this.#scrollableElement?.removeEventListener("scroll", this.#onScroll);
+      this.#scrollableElement = el;
+      this.#scrollableElement?.addEventListener("scroll", this.#onScroll);
+    }
+  }
+
+  #findScrollableParent(): HTMLElement | null {
+    if (!this.auto_scroll) return null;
+
+    // eslint-disable-next-line
+    let el: HTMLElement | null = this;
+    while (el) {
+      if (el.scrollHeight > el.clientHeight) return el;
+      el = el.parentElement;
+    }
+    return null;
+  }
+
+  #maybeScrollToBottom(): void {
+    const el = this.#scrollableElement;
+    if (!el || this.#isUserScrolled) return;
+
+    el.scroll({
+      top: el.scrollHeight - el.clientHeight,
+      behavior: this.streaming ? "instant" : "smooth",
+    });
+  }
+
+  #cleanup(): void {
+    this.#scrollableElement?.removeEventListener("scroll", this.#onScroll);
+    this.#scrollableElement = null;
+    this.#isUserScrolled = false;
+  }
+}
+
+// ------- Register custom elements and shiny bindings ---------
+
+customElements.get("shiny-markdown-stream") ||
+  customElements.define("shiny-markdown-stream", MarkdownElement);
+
+function handleMessage(message: ContentMessage | IsStreamingMessage): void {
+  const el = document.getElementById(message.id) as MarkdownElement;
+
+  if (!el) {
+    showShinyClientMessage({
+      status: "error",
+      message: `Unable to handle MarkdownStream() message since element with id
+      ${message.id} wasn't found. Do you need to call .ui() (Express) or need a
+      output_markdown_stream('${message.id}') in the UI (Core)?`,
+    });
+    return;
+  }
+
+  if (isStreamingMessage(message)) {
+    el.streaming = message.isStreaming;
+    return;
+  }
+
+  if (message.operation === "replace") {
+    el.setAttribute("content", message.content);
+  } else if (message.operation === "append") {
+    const content = el.getAttribute("content");
+    el.setAttribute("content", content + message.content);
+  } else {
+    console.error(`Unknown operation: ${message.operation}`);
```

Comment:
Might as well throw since it's in a message handler and won't bring down the app. That would just increase visibility (in dev mode we'd see the error in the client console).